### ISSUE TITLE
Support for `beforeEach` and `afterEach`

### ISSUE THREAD

**Issue by pke** [lorenzofox3/zora#52]
Coming from `node-tap` which is way to slow for my taste I wonder if zora has support for `beforeEach` and `afterEach` patterns?

**Comment by lorenzofox3** [comment]
It is tempting but I consider such hooks anti patterns when in comes to testing, especially for unit testing(so do I for snapshots but it is another story).  
Such "features" will probably not make it in zora's core but sometimes [people make me change my mind](https://github.com/lorenzofox3/zora/issues/16) :) 

However zora is quite simple to extend, you can for example create a decorator:

```javascript
// beforeEach decorator
// name the function with 'zora_spec_fn' to make sure location is correct
const withBeforeEach = (spec) => function zora_spec_fn(assert) {
    const hooks = [];
    assert.beforeEach = fn => hooks.push(fn);
    const originalTest = assert.test.bind(assert);
    assert.test = function (...args) {

        const [description, specFn, ...rest] = args;

        for (const h of hooks) {
            h();
        }

        // we decorate specFn too, so beforeEach can be used recursively in nested tests as well without calling the decorator
        return originalTest(description, withBeforeEach(specFn), ...rest);
    };

    return spec(assert);
};
```

to be used as 

```javascript
//src code to test
const sum = (a, b) => a + b

test('some test', withBeforeEach(t => {

    let input = 0;

    t.beforeEach(() => input++);

    t.test(`foo`, t => {
        t.eq(sum(input, 1), 2);
    });

    t.test(`another foo`, t => {
        t.eq(sum(input, 1), 2); // fails (should be 3) to test whether "at" location is correct
    });

    t.test('nested', t => {
        let input = 0;

        t.beforeEach(() => input++);

        t.test(`foo`, t => {
            t.eq(sum(input, 1), 2);
        });

        t.test(`another foo`, t => {
            t.eq(sum(input, 1), 2); // fails (should be 3) to test whether "at" location is correct
        });
    });
}));
```
A compromise would be to ship such helper functions with zora or within another package

**Comment by pke** [comment]
I am cool with that. I have the same philosophy and never understood the reason behind snapshots. Thanks for showing the way to implement such. Maybe we can just put it in the wiki here as a recipe instead of creating a new package?

**Comment by jsdevtom** [comment]
@lorenzofox3 Hats off for not only keeping the API minimal, but providing a solution for the people, who need to use `beforeEach`. Thank you for your great work!

**Comment by jsdevtom** [comment]
For those looking for a TypeScript version:

> #### `with-before-each.types.ts`
> ```ts
>import type { Assert } from 'zora';
>
>export interface AssertWithBeforeEach extends Assert {
>  beforeEach(fn: () => unknown): void;
>}
>
>export type SpecWithBeforeEach = (t: AssertWithBeforeEach) => unknown;
> ```

> #### `with-before-each.ts`
>```ts
>import type { Assert, SpecFunction, TestFunction } from 'zora';
>import type { AssertWithBeforeEach, SpecWithBeforeEach } from './with-before-each.types';
>
>export function withBeforeEach(spec: SpecWithBeforeEach): SpecFunction {
>  return ((assert: AssertWithBeforeEach) => {
>    const hooks: Array<() => unknown> = [];
>
>    assert.beforeEach = (fn: () => unknown) => {
>      hooks.push(fn);
>    };
>
>    const originalTest = assert.test.bind(assert);
>
>    assert.test = function(...args: Parameters<TestFunction>) {
>
>      const [description, specFn, ...rest] = args;
>
>      for (const hook of hooks) {
>        hook();
>      }
>
>      // we decorate specFn too, so beforeEach can be used recursively in nested tests as well without calling the decorator
>      return originalTest(
>        description,
>        withBeforeEach(specFn) as (assert: Assert) => unknown,
>        ...rest,
>      );
>    };
>
>    return spec(assert);
>  }) as SpecFunction;
>}

```

**Comment by lorenzofox3** [comment]
Thanks for the TS version

**Comment by Alexey-Sch** [comment]
As far as I understand, the code snippet above does not allow using asynchronous beforeEach methods. Simply adding ```async/await``` to the ```hook()``` call doesn't solve the issue.

**Comment by lorenzofox3** [comment]
@Alexey-Sch yes obviously. 

If you need async hook you might write something similar to 

```javascript
const withBeforeEach = (spec) => function zora_spec_fn(assert) {
    const hooks = [];
    assert.beforeEach = fn => hooks.push(async () => fn());
    const originalTest = assert.test.bind(assert);
    assert.test = async function (...args) {

        const [description, specFn, ...rest] = args;

        for (const h of hooks) {
            await h();
        }

        // we decorate specFn too, so beforeEach can be used recursively in nested tests as well without calling the decorator
        return originalTest(description, withBeforeEach(specFn), ...rest);
    };

    return spec(assert);
};
```

Have not tried though

**Comment by Alexey-Sch** [comment]
@lorenzofox3 Yes, I tried this already, the problem is that the tests stop running and crash with the following error:

```
VM40 warning.js:41 (node:8172) UnhandledPromiseRejectionWarning: Error: test "init object"
tried to collect an assertion after it has run to its completion.
You might have forgotten to wait for an asynchronous task to complete
------
function zora_spec_fn(assert) {
    const hooks = [];
    assert.beforeEach = fn => hooks.push(async () => fn());
    const originalTest = assert.test.bind(assert);
    assert.test = async function (...args) {
        const [description, specFn, ...rest] = args;
        for (const h of hooks) {
            await h();
        }
        // we decorate specFn too, so beforeEach can be used recursively in nested tests as well without calling the decorator
        return originalTest(description, withBeforeEach(specFn), ...rest);
    };
    return spec(assert);
}

    at collect (node_modules\zora\dist\bundle\index.js:167:19)
    at test (node_modules\zora\dist\bundle\index.js:411:9)
    at Object.test (node_modules\zora\dist\bundle\index.js:422:20)
    at Object.assert.test (tests\with-before-each.ts:17:16)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
...

VM40 warning.js:41 (node:8172) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

VM40 warning.js:41 (node:8172) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I tried to investigate the problem. It turns out that this line of code: https://github.com/lorenzofox3/zora/blob/master/src/test.ts#L33 is executed later than this one: https://github.com/lorenzofox3/zora/blob/master/src/test.ts#L11

**Comment by lorenzofox3** [comment]
would that work ? 

```Javascript
import {test} from 'zora';

const wait = (time = 100) => new Promise(resolve => setTimeout(resolve, time));

const withBeforeEach = (spec) => function zora_spec_fn(assert) {
    const hooks = [];
    assert.beforeEach = fn => hooks.push(async () => fn());
    const originalTest = assert.test.bind(assert);
    
    assert.test = async function (...args) {
        const [description, specFn] = args;
        return originalTest(description, withBeforeEach(async (t) => {
            for (const h of hooks) {
                await h();
            }
            return specFn(t);
        }));
    };
    
    return spec(assert);
};

test(`with async hook`, withBeforeEach((t) => {
    let counter = 0;
    
    t.beforeEach(() => wait().then(() => {
        counter++;
    }));
    
    t.test('first one', (t) => {
        t.eq(counter, 1);
    });
    
    t.test('second one', (t) => {
        t.eq(counter, 2);
    });
}));
```

**Comment by Alexey-Sch** [comment]
The test runs correctly, but I found nuances in how Zora works. I suggest going here: https://github.com/lorenzofox3/zora/issues/71

**Comment by Alexey-Sch** [comment]
I slightly changed the code, did this:

```javascript
const withBeforeAfterEach = (spec) => function zora_spec_fn(assert) {
    const hooks = [];

    assert.beforeEach = fn => hooksBefore.push(async () => fn());
    assert.afterEach = fn => hooksAfter.push(async () => fn());

    const originalTest = assert.test.bind(assert);
    
    assert.test = async function (...args) {
        const [description, specFn] = args;
        return originalTest(description, withBeforeAfterEach(async (t) => {
            for (const h of hooksBefore) {
                await h();
            }

            const result = specFn(t);

            for (const h of hooksAfter) {
                await h();
            }

            return result;
        }));
    };
    
    return spec(assert);
};
```

In Jest, tests are executed sequentially. As I understand it, specifically so that common variables between tests can be reused. That is, something like this:

```javascript
const mocked = new UsedObjectMock();

const obj = new MyObject({
    UsedObject: mocked
});

beforeEach(() => {
    obj.init();
    console.log('beforeEach');
});

afterEach(() => {
    obj.release();
    console.log('afterEach');
});

test('test1', async () => {
    await mocked.setState(1);
    expect(obj.data()).toEqual(true);
    console.log('test1');
});

test('test2', async () => {
    await mocked.setState(2);
    expect(obj.data2()).toEqual(true);
    console.log('test2');
});
```

```
> beforeEach
> test1
> afterEach
> beforeEach
> test2
> afterEach
```

In the case of Zora, this approach will call ```obj.init()``` twice in a row (since all tests are executed in parallel), which will negatively affect the passing of tests:

> beforeEach
> beforeEach
> test1
> afterEach
> test2
> afterEach

Am I getting it right? Is there any optimal solution here?

**Comment by lorenzofox3** [comment]
To be honest I don't know, I have not dug too much. The hooks I mentioned in this thread are just examples to consider as so, and are not part of zora.

I personally find the usage of such constructs a bad practice (that is why they are not part of the core library) and would rather consider each test independently.

```javascript
test(`some test`, () => {
    prepare();

    // do some expectations

   teardown()
})

test(`some other test`, () => {
    prepare();

    // do some expectations

   teardown()
})
```

Which are more explicit and easier to maintain in the long term. I would eventually factorize with a spec decorator to follow DRY principles (like any code) but certainly not use things like ``beforeEach`` which create global context and tend to introduce coupling between the tests. 

That being said. You are correct: by default zora run all the spec function concurrently. However you can explicitly maintain a shared state or a sequence between sub tests with the ``async/await`` control flow like would to for any other JS code:

```Javascript

test(`some test suite`, async (t) => {
   let state = 0;

   //explicitly wait for the test to complete before moving on
  await t.test(`sub test 1`,async (t) => {
       await wait();
       t.eq(state, 0);
       state++;
  });

  await t.test(`sub test 2`,async (t) => {
       await wait();
       t.eq(state, 1);
       state++;
  });

  // etc

})

```

**Comment by Alexey-Sch** [comment]
Unfortunately, the parallel test execution, which is common in `zora`, is not suitable for all projects.

Often you have to test asynchronous code with this package: https://github.com/sinonjs/fake-timers (Jest uses its own implementation: https://jestjs.io/docs/en/timer-mocks)

At the beginning of the test, it overwrites the `setTimeout`, `setInterval`, and so on methods with its own implementation, and at the end of the test it replaces the native implementation of these methods.

And if inside one file with a set of tests it is still possible to achieve sequential execution of tests by placing the await keywords, then in the case of the `pta` package it runs several files in parallel and they begin to conflict with each other. Is there a way to avoid this?

Most likely this question should have been asked in the `pta` repository, but this issue contains the whole discussion context, so I ask this question here.

**Comment by Alexey-Sch** [comment]
I solved this problem by adding `await` in this line: https://github.com/lorenzofox3/zora-node/blob/master/src/cli.js#L84

In general, it turned out that for my needs it is easier to rewrite the file (since it is small in size) than to install the `pta` package.
